### ISSUE TITLE
Automatic generation of API documentation

### ISSUE THREAD

**Issue by mrbean-bremen** [fo-dicom/fo-dicom#1199]
Currently, there only seems to exist a [static documentation repository](https://github.com/fo-dicom/fo-dicom.github.io), which seems to be pretty much frozen. It would be nice to have automatic API documentation generation in version 5, which could be published on GitHub Pages, ReadTheDocs or elsewhere.

One possibility is to use DocFX - for a simple example see the [documentation of SVG.NET](http://svg-net.github.io/SVG/), which includes the [API documentation](http://svg-net.github.io/SVG/api/Svg.html), and is created by a [GitHub workflow](https://github.com/svg-net/SVG/blob/master/.github/workflows/builddocs.yml).

This could also be a place to add example code, FAQs etc. in the future.

**Comment by mrbean-bremen** [comment]
Actually, I didn't notice that there is already some setup for documentation generation using the Sandcastle Tools, which probably has been used to generate the static pages. This can be used instead of DocFX, but still needs to be integrated into a GitHub workflow.

**Comment by gofal** [comment]
Anders Gustafsson set up the sandcastle project because it had some advantages over other tools. But I am not sure if sandcastle is still alive.
One important thing about the tool is, that it can be generated from dotnet xml documentation comments. I think the integration into guthub workflow is a nice-to-have, but not required. Because I am releasing a nuget-package every few months, which requires some manual steps. updating the static documentation would just be one more step. I think the api documentation should rather be in sync with the latest nuget package than with the current development branch on github?

Do you have experiences with any of the documentation generation tools?

**Comment by mrbean-bremen** [comment]
I don't have much experience, but I wrote the workflow with DocFX in SVG.NET that I linked above. DocFX does the same (e.g. generate documentation from the XML comments) as Sandcastle (which I didn't know about).
What I do in other (Python) projects is to generate the documentation for master (that would be development in fo-dicom) automatically, and copy it over to the release documentation on release.

**Comment by mrbean-bremen** [comment]
Sandcastle is still in active development (not the original Microsoft tool, but the [Sandcastle Help File Builder](https://github.com/EWSoftware/SHFB)), and doesn't seem to be too different from DocFX regarding the actual documentation generation. It uses another (XML-based) markup language (DocFX use Markdown), and the configuration file is XML instead of JSON, and seems to have more features. 

Anyway, if you think this is not needed, you can just close the issue - it was just an idea (I like to automate stuff, as I tend to forget things...)

**Comment by gofal** [comment]
> Anyway, if you think this is not needed, you can just close the issue - it was just an idea (I like to automate stuff, as I tend to forget things...)

Everything that is helpful is needed! Documentation is one of the aspects that is really neglected in this project. So would be great if you would go and push this forward. 👍

**Comment by mrbean-bremen** [comment]
Ok, I will see what I can do - no promises regarding the timeline, it may take a while before I get to it!

**Comment by mrbean-bremen** [comment]
I had a closer look at the Sandcastle Help File Builder, and while it works fine, it is difficult to install and very heavy-weight, and I don't see how to easily integrate it into CI. DocFX is much easier to handle, can be installed via choco, and also generates readable API documentation, though arguably less colorful than Sandcastle. OTOH, the documentation you write (apart from the generated API doc) is in a readable markup format, while Sandcastle uses the more obscure (IMHO) MAML language (Microsoft Assistance Markup Language) which is difficult to create (but may have more features).
So I will probably try to use DocFX and check if the result is sufficient.

**Comment by gofal** [comment]
cool, thanks! The best tool with the best featues is useless if it is then so difficult to install or maintain that in the end noone uses it...

**Comment by mrbean-bremen** [comment]
I have been playing around with DocFX a bit, and still don't know how to do proper navigation, but you may check of it how it looks [here](https://mrbean-bremen.github.io/fo-dicom/dev/index.html). A few notes:
- apart form the layout/navigation this also lacks content - I put there a small part of the usage notes, as I wasn't sure that the old notes are still valid; I may need some help here 
- the linked generated API documentation will probably not be changed (apart from maybe some styling changes), please check if that is acceptable
- the icon on the upper left is for navigation to the home page, in case that is not obvious
- I push the changes to the `gh-pages` branch of my own fork, as was the default with GitHub Pages; to make this to work in the original repo someone needs to create a `gh-pages` branch and clean it (e.g. remove all files, as only the documentation will live here), and enable GitHub Pages for this branch in the settings (under `Pages`); this was just the easiest way to do it
- the documentation is commited into the `dev` path with the idea to also have a `stable` path (or more paths for different releases), where the dev part is copied on release

**Comment by mrbean-bremen** [comment]
Ah, and now that I see the Wiki documentation that I had completely missed, this can be linked or moved over to the documentation, too.

**Comment by gofal** [comment]
Looks fine. And it looks very familar. I have seen api documentations like that already many times.
I will do the final version 5 release now (working on that with Efferent) and then drop everything from the readme files that are related to version 4 and prior. So this may change.
Is it a manual work, to include those notes, or are the readme files imported automatically? The same with the wiki, would it then be better to update the documentation or the wiki? can docxfx import the content of the wiki automatically?

I will check that with the pages and branch.

**Comment by mrbean-bremen** [comment]
The md files in the repository are referenced and imported automatically, not sure about the Wiki, have to check that. But much of the Wiki looks outdated, so maybe it makes sense to move it to the documentation in the repository, so that it will be possible to make version-specific documentation.
I can just make draft PR with the current changes made for the documentation, so it is better visible.

**Comment by mrbean-bremen** [comment]
Ok, the Wiki files can only be linked  to - we don't really have access to the content there.

**Comment by mrbean-bremen** [comment]
I fixed the navigation, now the question is mostly about content. 
I'm not sure about the Wiki pages - we can just move them over to the documentation (e.g. add them to the repository), link them in the documentation (I did this for one page as an example), or just leave it completely separate. What do you think?
Also, is there more existing documentation that shall be added, or some introduction needed?
We can always more documentation later, though.

**Comment by mrbean-bremen** [comment]
Maybe we can just use the README as the intro page, or at least some information from there.

**Comment by gofal** [comment]
Yes, the content is a huge question. A lot of the content is outdated. Now there are also two versions (4 and 5) that do have slightly different API.
I have been hesitant and propably too careful with releasing version 5, but after over a year running version 5 fine in production (at least in our company) I now will to the release. I released the 4.0.8 containing the last bugfixes in version 4. And now I will do a version 5 release. Then I will update the content and drop outdated stuff and update the code samples to match version 5. 
I am affraid this will take some time, but I would like to do it in sync with the documentation generation. 

But I am unsure if its better to add content to github-wiki or to a documentation page. One thing is for sure: there should only be one place to avoid double work or inconsistent contents. On questions I will point the devs to the place where to find the content, so I'd prefer the place, where it is easier to maintain. 
Content in github-wiki can be changed very easily. But content in generated documentation can be changed within visual studio while developing and can also be updated by contributors in pull requests.

**Comment by mrbean-bremen** [comment]
> But I am unsure if its better to add content to github-wiki or to a documentation page.

I'm also unsure, but in other (Python) projects I tend to put all documentation in one place and link to the place prominently on the homepage. Adapting existing documentation should be relatively easy (though not as easy as in the Wiki pages), because the documentation is saved as markdown files (.md) same as README.md and other files, most developers should be familiar with that. But I leave that decision to you, of course.
There will be a bit of duplication with the README, but we could also move some of the README into the documentaion pages.

As for version 4: It may make sense to also generate documentation for that version. If development will proceed for both versions for a while, we could generate separate documentation for version 4 and put that in a separate path (instead of the `dev` branch we could have `dev4` and `dev5` branches, and corresponding `stable4` / `stable5` paths, where the documentation is copied on release).

**Comment by amoerie** [comment]
Thanks for pinging me @gofal. I think just mentioning me in an issue with @amoerie also sends me an email, so that works too. 🙂

I've read the discussion so far, and these are some of my thoughts:

- If we have "API docs", they should be generated or removed. I think step 1 should be removing the outdated docs, because the only thing worse than no documentation is wrong documentation.
- If we generate API docs, there should be some way of asking for the docs of a specific version. Especially with our special "v5" and "v4" story.
- Manually written documentation is always more valuable than generated docs. Simple guides such as "How to open and modify a DICOM file", "How to render an image" and "How to send a DICOM request" are infinitely more useful than the generated DicomFile.cs or DicomClient.cs API docs. Manually written docs are more important and should be more prominently featured in our readme. 
- Manually written documentation should be source controlled (in .MD files) so we have a historically accurate version of manual documentation
- Generated documentation can be source controlled but I'd rather not, since this will massively increase the number of files we have in GIT. I'd like to avoid it if we can.

**Comment by mrbean-bremen** [comment]
Thanks @amoerie - looks like we are mostly on the same page here.

> If we generate API docs, there should be some way of asking for the docs of a specific version.

Yes - this could either be done in a landing page that just points to the specific doumentation, or in separate links in the README (or both). Technically, this could be done by using separate subdirectories, as mentioned.

> Manually written documentation is always more valuable than generated docs.

Certainly. The point is writing it, of course... Also note that this issue mixes these up a bit, as the [generated pages](https://mrbean-bremen.github.io/fo-dicom/dev/index.html) would include both API documentation and written documentation, the idea is just to have one place for all documentation. The generated API documentation will just be one entry in the documentation.

> Manually written documentation should be source controlled (in .MD files) so we have a historically accurate version of manual documentation.

Agreed. This is my problem with Wiki pages - they are fine for documentation not related to a specific version, but otherwise they tend to get outdated quite fast (as can be seen with the current Wiki in fo-dicom). If they are under version control, it would usually be caught in the review if the pages have not been updated. Probably should adapt the PR check list to include this...

> Generated documentation can be source controlled but I'd rather not, since this will massively increase the number of files we have in GIT. I'd like to avoid it if we can.

Hm, that is a problem with the current approach (e.g. the `gh-pages` branch). Currently the generated API documentation will be checked into that branch, which is used to serve the documentation page (the development branches are not affected, but the repository size itself _will_ increase due to that branch). We could use the separate documentation repository instead (which currently has documentation for version 3), given the correct permissions it should be possible to push to another repository.

**Comment by amoerie** [comment]
I'm not a difficult person, and files in GIT don't cost money. If there is a convincing reason to have them here, we should do it. But if there is a (very) easy path to store them outside of GIT, I would like to consider it.

Furthermore, is there demand for this generated API documentation? 
I personally don't use it, since the .NET IDEs (Visual Studio / JetBrains Rider) are very capable of navigating DLL + PDB files, and they allow you to even read the (sometimes decompiled) source code in one fell swoop.

**Comment by mrbean-bremen** [comment]
> I'm not a difficult person, 

😁 

> If there is a convincing reason to have them here, we should do it. But if there is a (very) easy path to store them outside of GIT, I would like to consider it.

There are certainly other possibilities out there, like ReadTheDocs, I'm just not familiar with them. But putting the documentation into a separate repository should not be a big problem, I think, as usually this repository will not needed to be cloned locally (well, if we want to copy documentation for a release, but that could be automated, too).

> Furthermore, is there demand for this generated API documentation?

Maybe not... It is more or less standard to provide this, and there is currently the documentation for version 3 (which actually prompted me to do this), but I personally don't need it (I'm ok with what VS/Rider are providing). I have no idea how many people are actually relying on that.

Even if we don't need that, I still think it would be a good idea to have generated documentation pages from the manually written documentation, as it allows for easier navigation and overview.

**Comment by mrbean-bremen** [comment]
And I have no problem if we just close this PR if it is not needed, but I still think moving the documentation from the Wiki into the repository would be a good idea anyway.

**Comment by amoerie** [comment]
Well if the demand is unknown or not very high, that doesn't mean we shouldn't do it. But it also means we probably shouldn't spend a lot of effort. What's the easiest way then to have auto-generated API docs per version?

**Comment by amoerie** [comment]
> And I have no problem if we just close this PR if it is not needed, but I still think moving the documentation from the Wiki into the repository would be a good idea anyway.

Agreed

**Comment by mrbean-bremen** [comment]
> What's the easiest way then to have auto-generated API docs per version? A subfolder in this repository that gets generated by a script?

I would say separate docbuild workflows for version 4 and 5 with different inputs. Basically copy the existing one with different inputs and a different output folder. We also should have different versions for the manually created documentation in this case, at least where it will differ.

**Comment by amoerie** [comment]
OK, that sounds good to me. 

Is this TODO list a correct interpretation? Please correct me if I'm wrong

- Move wiki to markdown files in repo. Duplicate wiki for v4 and v5, and adapt both version so that they are correct for each version. v4 is in "maintenance mode" so it's ok if these docs are more incomplete.
- Setup doc build script so we can generate API docs both for v4 and v5. For simplicity, store the result in this git repo too. That allows the API docs to remain in sync with releases easily. We only have to remember to run this script every release. 
- Include our manually written documentation in the DocFX output somehow
- Add link in README to the latest generated docs for v4 and v5.

Open question: if the results are stored in this repository, is the doc script something that can be automated using Github Actions? Because it would need to commit changes, presumably. Or am I missing something?

**Comment by mrbean-bremen** [comment]
Sounds good! 

> if the results are stored in this repository, is the doc script something that can be automated using Github Actions? Because it would need to commit changes, presumably.

This is what this PR is about. The GitHub workflow is triggered automatically on push (only in my fork at the moment), the [result](https://mrbean-bremen.github.io/fo-dicom/dev/index.html) has been generated automatically. I have only included a few self-written manuals (due to the lack of current ones...) to show how it could look.

**Comment by amoerie** [comment]
Oh I missed you had already created a PR. Perfect!

**Comment by mrbean-bremen** [comment]
> Are you willing to finish it up then with the things we concluded?

Sure! I will wait for @gofal to clear up the documentation. I can do the move from the Wiki after it is clear what shall remain, and add separate workflows for versions 4 and 5.

**Comment by gofal** [comment]
This sounds perfect for me. I Also have released version 5.0.0 (no prerelease-postfix) today. My next step is now to update the current manual documentation to match to version5.
So I will not do that in github wiki now, but wait that you migrated the pages into the repository and duplicated it for version 5, then I will update them.

**Comment by mrbean-bremen** [comment]
Ok, I will see if I can get to it tonight, or in the near future.

**Comment by gofal** [comment]
About publishing the documentation. I try to sum it up, correct me if I missed something.
fo-dicom organization has a separate repository fo-dicom.github.io. There I will copy the current content of the generated documentation on every release manually.
There is this Pages-feature in every github project, where I can specify a directory, whose content will be served like from a webserver. So the documentation will be generated into a directory within fo-dicom repository automatically on every push. This directory will be shared with this Pages-feature and therefore the current documentation is also available.

**Comment by mrbean-bremen** [comment]
Yes, that basically is it. Manually copying the contents to the other repository is one possibility, you could also use the same repository, but different subdirs. We could also add a GitHub workflow that does this automatically on release (for example based on the release branch name).
I may also check if I can commit all documentation into a separate repository (e.g. fo-dicom.github.io). I have not done this before, so I'm not sure if this is possible, but I can just fork the docu repository and try this.

**Comment by mrbean-bremen** [comment]
@gofal - I have copied over most (not all) of the documentation from the Wiki in the PR, basically copied them twice into the version 4 and 5 paths. I didn't do any version specific adaptations (I understood that you were going to do this), and I haven't tried to use the other repository yet. The generated documentation pages for [version 4](https://mrbean-bremen.github.io/fo-dicom/dev/v4/index.html) and [version 5](https://mrbean-bremen.github.io/fo-dicom/dev/v5/index.html) are mostly similar at this point (I also didn't add a link to the documentation yet).
I will try to get that to a point where it can be merged somewhere in the next days.

**Comment by mrbean-bremen** [comment]
If you want, you can already adapt some of the documentation (mostly under Documentation/v5/usage) in the PR branch.

**Comment by gofal** [comment]
thanks!

**Comment by mrbean-bremen** [comment]
I have now changed the build script to push to the `fo-dicom.github.io` repository to avoid enlarging the `fo-dicom` repository, as discussed with @amoerie. To make this work, the `fo-dicom.github.io` repo has to be prepared by adding the directories `dev/v4` and `dev/v5` - I didn't add this in the script, as it has to be done only once. The question is if we want to retain the documentation for fo-dicom 3. In my fork, I have deleted all files, but we can also leave it all, as the new new documentation will reside in separate subdirectories. This is probably better - we can always remove the files later. What do you think? I will add a PR to add the directories once you agree that this is the way to go.

Also, to make it possible to push to another repo, we need a GitHub token to allow this. Someone with admin rights in fo-dicom has to create an API token with "repo" access, and then add this token as a repositiory secret to the fo-dicom repository under the name `API_TOKEN_GITHUB` (I use this name in the script). Once this is done, I can change the PR to use the main repo instead of my fork, and can enable the script for pull requests temporarily to be able to test this in the PR. 

I have not touched the documentation itself, I think this can be done separately. Let me know what I have to adapt in the PR. The currently created documentation can be found at https://mrbean-bremen.github.io/fo-dicom.github.io/dev/v4/ and https://mrbean-bremen.github.io/fo-dicom.github.io/dev/v5/.

**Comment by gofal** [comment]
I started to update doc to version 5. The samples repository is now on version 5 and I updated the main readme.me file of the fo-dicom repository. Now I'd like to edit the documentation files.
I love to work within visual studio - i do also have a plugin that renders the md textfile while I edit it and shows a preview. And if the md file are included in visual studio solution then that reminds me on edit those files too when doing code changes.
Do you have the md files visible in visual studio? Should I create some pseudo-project that does nothing but only has the linked md files as content?

**Comment by mrbean-bremen** [comment]
I don't have them in the project, but I think it's definitely a good idea. Go ahead, of course!

**Comment by mrbean-bremen** [comment]
I think we can close this, as the automatic API documentation generation works now. The rest of the work is to improve/fix the other documentation.
